### ISSUE TITLE
Windows x64 performance issue, high number of exceptions

### ISSUE THREAD

**Issue by DongYeopKang** [LuaJIT/LuaJIT#781]
Hello!
I developed a 32bit game using c++ and luajit 2.1
I'm currently trying to port to 64-bit, but frame performance drops a lot (about 30-50%)
It is a structure that calls a lot of lua_pcall every frame, but when profiling, the frame drops were severe in this part.
The first few lua_pcall are fast, but sometimes lua_pcall calls take a long time in between.
Most lua_pcalls are less than 0.002ms, but sometimes it takes more than 3ms in between.
All lua_pcall arguments and calling methods are the same.
Numerous kernel exceptions are also occurring, so when I looked up the stack, countless errors were occurring in the process of mcode_alloc through lua_pcall. (KernelBase exception 0xE24C4A02)
Is this an issue with gc64?
Please let me know if you need more information.
I have been suffering from this problem for a long time and I really need your help.

**Comment by MikePall** [comment]
Closing due to lack of a follow-up from your side.

**Comment by hrsantiago** [comment]
Hello, I'm having a similar problem to this one.
Also had a 32 bit game and just ported it to Windows 64 bit.

It works just fine for most users (maybe 95%+), but a few of them gets random crashes always on LuaJIT.
A few backtraces for context:

```
0: C:\Windows\System32\KERNELBASE.dll [0x00000000398ecd29] Unknown
1: C:\myapp.exe [0x0000000000de27ac] lj_err_throw
2: C:\myapp.exe [0x0000000000de7a9d] lj_str_new
3: C:\myapp.exe [0x0000000000de7af1] lj_str_new
4: C:\myapp.exe [0x0000000000e0bcb4] lj_tab_reasize
5: C:\myapp.exe [0x0000000000e0bd79] lj_tab_new
6: C:\myapp.exe [0x0000000000e2e43e] lua_createtable
7: C:\myapp.exe [0x00000000007a0fed] push_luavalue(LuaInterface*, TRect<int> const&)
8: C:\myapp.exe [0x0000000000872d67] bool emit<stdext::shared_object_ptr<LuaObject>, char [17], TRect<int>, TRect<int> >(stdext::shared_object_ptr<LuaObject> const&, char const (&) [17], TRect<int> const&, TRect<int> const&) [clone .constprop.0]
9: C:\myapp.exe [0x00000000004ae2b6] UIWidget::onGeometryChange(TRect<int> const&, TRect<int> const&)
10: C:myapp.exe [0x000000000077b785] Event::execute()
11: C:myapp.exe [0x00000000008abbbf] EventDispatcher::poll() [clone .constprop.0]
12: C:myapp.exe [0x00000000005de186] GraphicalApplication::poll()
13: C:myapp.exe [0x00000000005e1013] GraphicalApplication::frame()
14: C:myapp.exe [0x000000000078dd79] Application::run()
15: C:myapp.exe [0x0000000000f18414] main
16: C:myapp.exe [0x00000000004013c1] 0x520 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/../../../../x86_64-w64-mingw32/lib/crt2.o
17: C:myapp.exe [0x00000000004014d6] WinMainCRTStartup
18: C:\Windows\System32\KERNEL32.DLL [0x000000003b3f74b4] Unknown
19: C:\Windows\SYSTEM32\ntdll.dll [0x000000003c2226a1] Unknown


0: C:\WINDOWS\System32\KERNELBASE.dll [0x000000008172cd29] Unknown
1: C:\myapp.exe [0x0000000000de27ac] lj_err_throw
2: C:\myapp.exe [0x0000000000de7a9d] lj_str_new
3: C:\myapp.exe [0x0000000000de7af1] lj_str_new
4: C:\myapp.exe [0x0000000000e2ec1f] lua_newuserdata
5: C:\myapp.exe [0x00000000007a4a47] LuaInterface::pushObject(stdext::shared_object_ptr<LuaObject> const&)
6: C:\myapp.exe [0x0000000000872d36] bool emit<stdext::shared_object_ptr<LuaObject>, char [17], TRect<int>, TRect<int> >(stdext::shared_object_ptr<LuaObject> const&, char const (&) [17], TRect<int> const&, TRect<int> const&) [clone .constprop.0]
7: C:\myapp.exe [0x00000000004ae2b6] UIWidget::onGeometryChange(TRect<int> const&, TRect<int> const&)
8: C:\myapp.exe [0x000000000077b785] Event::execute()
9: C:\myapp.exe [0x00000000008abbbf] EventDispatcher::poll() [clone .constprop.0]
10: C:myapp.exe [0x00000000005de186] GraphicalApplication::poll()
11: C:myapp.exe [0x00000000005e1013] GraphicalApplication::frame()
12: C:myapp.exe [0x000000000078dd79] Application::run()
13: C:myapp.exe [0x0000000000f18414] main
14: C:myapp.exe [0x00000000004013c1] 0x520 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/../../../../x86_64-w64-mingw32/lib/crt2.o
15: C:myapp.exe [0x00000000004014d6] WinMainCRTStartup
16: C:\WINDOWS\System32\KERNEL32.DLL [0x0000000081b074b4] Unknown
17: C:\WINDOWS\SYSTEM32\ntdll.dll [0x0000000083a426a1] Unknown


0: C:\Windows\System32\KERNELBASE.dll [0x00000000398ecd29] Unknown
1: C:\myapp.exe [0x0000000000de27ac] lj_err_throw
2: C:\myapp.exe [0x0000000000de7a9d] lj_str_new
3: C:\myapp.exe [0x0000000000dfb04a] lj_buf_putstr_upper
4: C:\myapp.exe [0x0000000000e2e0d7] lua_pushcclosure
5: C:\myapp.exe [0x00000000007a6cec] LuaInterface::luaCall(int, int)
6: C:\myapp.exe [0x00000000008aebed] bool LuaInterface::callGlobal<bool, stdext::shared_object_ptr<LuaObject>, char [11]>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, stdext::shared_object_ptr<LuaObject> const&, char const (&) [11]) [clone .constprop.0] [clone .isra.0]
7: C:\myapp.exe [0x0000000000420199] bool emit<stdext::shared_object_ptr<LuaObject>, char [15]>(stdext::shared_object_ptr<LuaObject> const&, char const (&) [15])
8: C:\myapp.exe [0x00000000004ad820] UIWidget::onLayoutUpdate()
9: C:\myapp.exe [0x000000000077b785] Event::execute()
10: C:myapp.exe [0x00000000008abbbf] EventDispatcher::poll() [clone .constprop.0]
11: C:myapp.exe [0x00000000005de186] GraphicalApplication::poll()
12: C:myapp.exe [0x00000000005e1013] GraphicalApplication::frame()
13: C:myapp.exe [0x000000000078dd79] Application::run()
14: C:myapp.exe [0x0000000000f18414] main
15: C:myapp.exe [0x00000000004013c1] 0x520 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/../../../../x86_64-w64-mingw32/lib/crt2.o
16: C:myapp.exe [0x00000000004014d6] WinMainCRTStartup
17: C:\Windows\System32\KERNEL32.DLL [0x000000003b3f74b4] Unknown
18: C:\Windows\SYSTEM32\ntdll.dll [0x000000003c2226a1] Unknown


0: C:\WINDOWS\System32\KERNELBASE.dll [0x000000003993cd29] Unknown
1: C:\myapp.exe [0x0000000000de1d6a] lj_vm_log2
2: C:\myapp.exe [0x0000000000de6cae] lj_func_closeuv
3: C:\myapp.exe [0x0000000000de6d02] lj_meta_equal
4: C:\myapp.exe [0x0000000000de6b51] lj_func_closeuv
5: C:\myapp.exe [0x0000000000e25a0c] lj_meta_arith
6: C:\myapp.exe [0x00000000008b332a] bool emit<char [9], char [6], LogLevel, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned long long>(char const (&) [9], char const (&) [6], LogLevel const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long long const&) [clone .constprop.0] [clone .isra.0]
7: C:\myapp.exe [0x000000000077b785] Event::execute()
8: C:\myapp.exe [0x00000000008abbbf] EventDispatcher::poll() [clone .constprop.0]
9: C:\myapp.exe [0x00000000005de186] GraphicalApplication::poll()
10: C:myapp.exe [0x00000000005e1013] GraphicalApplication::frame()
11: C:myapp.exe [0x000000000078dd79] Application::run()
12: C:myapp.exe [0x0000000000f11b94] main
13: C:myapp.exe [0x00000000004013c1] 0x520 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/../../../../x86_64-w64-mingw32/lib/crt2.o
14: C:myapp.exe [0x00000000004014d6] WinMainCRTStartup
15: C:\WINDOWS\System32\KERNEL32.DLL [0x000000003a517034] Unknown
16: C:\WINDOWS\SYSTEM32\ntdll.dll [0x000000003c2626a1] Unknown

```

Most of the crashes ends with lj_err_throw and lj_str_new
Looking at the LuaJIT code that it was failing to allocate memory, and that's how I got on this issue.

I've found one user that can reproduce the crash almost always when starting the application. And could test the 

```
require("jit.v").start("dump.txt")
```

His output is the following (I've truncated it to be smaller, but it just repeats)

```
...
[TRACE --- (182/1) uihtml.klua:0 -- NYI: return to lower frame]
[TRACE --- uiwindow.klua:0 -- NYI: return to lower frame]
[TRACE --- (182/1) uihtml.klua:0 -- NYI: return to lower frame]
[TRACE --- iterators.klua:0 -- NYI: bytecode 51]
[TRACE 183 uitextedit.klua:0 return]
[TRACE 184 uimdiwindow.klua:0 return]
[TRACE 185 types.klua:0 return]
[TRACE 186 uiscrollbar.klua:0 return]
[TRACE 187 uiscrollbar.klua:0 stitch C:7a6c50]
[TRACE 188 (187/stitch) util.klua:0 -> 7]
[TRACE --- uicalendar.klua:0 -- leaving loop in root trace]
[TRACE 189 styles.klui:149: [@slot onTextChange]:1 return]
[TRACE 190 uicalendar.klua:0 return]
[TRACE --- uicalendar.klua:0 -- leaving loop in root trace]
[TRACE --- guarder.klua:0 -- NYI: bytecode 51]
[TRACE 191 uiscrollbar.klua:0 return]
[TRACE --- uitextedit.klua:0 -- failed to allocate mcode memory]
[TRACE flush]
[TRACE --- uitextedit.klua:0 -- failed to allocate mcode memory]
[TRACE flush]
[TRACE --- uitextedit.klua:0 -- failed to allocate mcode memory]
[TRACE flush]
[TRACE --- luaengine:291 -- failed to allocate mcode memory at luaengine:292]
[TRACE flush]
[TRACE --- uiwidget.klua:0 -- failed to allocate mcode memory]
...
```

"failed to allocate mcode memory" starts very often but probably it is not due to bad Lua code, as this doesn't happen on the Win32 bit version nor Linux version and Lua memory usages stays low the whole time.

I've took a look at issues https://github.com/LuaJIT/LuaJIT/issues/282, https://github.com/LuaJIT/LuaJIT/issues/283, https://github.com/LuaJIT/LuaJIT/issues/284, https://github.com/LuaJIT/LuaJIT/issues/285, but they don't seem really related, as they talk about multi thread and low memory.

 I'll keep digging into: "something in your system doesn't allow it to allocate executable memory within +-1GB of the LuaJIT DLL."

**Comment by MikePall** [comment]
It looks like they are actually running out of memory. The failed mcode memory allocation is just one symptom and is probably not the main issue.

I hope you're using the (default) GC64 mode when building LuaJIT v2.1? Please do not use old releases.

Some of those backtraces do not make sense. I guess the debug info doesn't match the binary? I can't see how you are calling into the VM. You need to use `lua_pcall()` or `lua_cpcall()`, otherwise you won't be able to catch an out-of-memory error.

If you insist on catching Lua errors via C++ `catch`, you need to compile all of your C++ code to catch foreign exceptions on Windows. For MSVC that's `/EHa`, don't know about MinGW. Please note that catching foreign exceptions has some side effects, like catching access violations, too.

Anyway, I don't think this is related to the original issue at all.

**Comment by hrsantiago** [comment]
The last backtrace was indeed wrong, sorry about that. I've removed it.
My problem was solved, but leaving this message here as it might help others.

I wasn't using GC64 because I still deploy the 32 bit client, as 3% of users still have Windows 32 bits.
Both 32bit and 64bit executables read the same Lua files generated with luajit -b, so that's why I was not using GC64.

Reading this: https://blog.openresty.com/en/luajit-gc64-mode/
It seems that my problem was that my app uses 1GB+ of non Lua memory and the system was lacking of low addresses to allocate Lua memory.
If I understand right, LuaJIT can only allocate memory below 2 GB addresses when compiled in 64 bit without GC64.

After deploying 2 versions of luajit -b files and recompiling LuaJIT with GC64 the crash no longer occurs.
